### ISSUE TITLE
Discussion: First class accommodations for multiple simultaneous instances of a single form

### ISSUE THREAD

**Issue by joshacheson** [davidkpiano/react-redux-form#868]
### The Problem

It doesn't appear to me that there's a clear solution / best practice for having multiple instances of a single form on the page at a given point in time. One solution (which would take place in user land) is to simply dynamically add more reducers / places in state in a given application, for instance if I have been keeping my forms in state like:

```javascript
{
  userProfileForm: FormReducer,
  changePasswordForm: FormReducer
}
```

and I wanted to have multiple user profile forms on the page simultaneously, I can dynamically add a reducer and end up with:

```javascript
{
  userProfileForm1: FormReducer,
  userProfileForm2: FormReducer
}
```

I think I could work something out in my own app to do this.

With that said, it might be nice to accommodate this concern in RRF itself (or maybe there's a way to do this now, and I've just not noticed it!).

I'm filing this issue to host a discussion about what the community/maintainer(s) thing about this topic, and if it's determined that RRF would like to handle this (as opposed to deferring solutions to user land) I would happily follow up and do the work to implement this feature.

**Comment by davidkpiano** [comment]
This should work naturally:

```jsx
// with this store...
const store = createStore(combineForms({
  allForms: []
});
```

```jsx
// and this in render()...
allForms.map((form, index) => {
  <Form model={`allForms[${index}]`}>
    <Control.text model=".foo" />
  </Form>

// somewhere else in render()...
<button onClick={() => dispatch(actions.push('allForms', {}))}
  Make new form
</button>
```

Is this what you mean?

**Comment by joshacheson** [comment]
This approach is one I hadn't considered. I'll give this a shot and report back with my findings here. In the mean time I will close this issue, and re-open for more discussion only if it's necessary. I really appreciate the quick response.

I'm not sure if there's a place in the docs that references this sort of usage, and I'd be happy to PR to the docs to add an entry under Recipes and Examples if you'd want something like that (again, assuming this solution pans out).

**Comment by daxiang28** [comment]
@joshacheson I was wondering if you were able to move past a simple proof of concept of the example defined here. A simple form seems to work in my case except for when field level validation is required.

**Comment by joshacheson** [comment]
I never ended up using global forms via `react-redux-form`. What we did instead was always use "local forms" (although not `react-redux-forms`'s `<LocalForm/>`, but instead manually created stores which we pass to `<Form/>`, so in effect a "local form") and we "register" those forms to a namespace on a custom `<FormsProvider/>` at the top of our tree component tree. This way, we get the benefits of local forms (performance, mainly, by not dispatching to the global redux store all the time) but still have the ability to read from / write to form state all around the application. Our abstractions on top of `react-redux-form` go well beyond this (we've also abstracted its actions API entirely) but I digress!

If you're interest in a solution like the one I described I'd be happy to help / describe it in more depth.

**Comment by daxiang28** [comment]
@joshacheson I appreciate the detailed response. I think we're headed down a similar route, but could probably get away with the RRF LocalForm since we don't have a lot of form instances. I am a little curious about the abstraction of the actions API and what you dit there.

**Comment by joshacheson** [comment]
Yeah! So there are a couple different things we did there:

1) We wrote a `Form` class (as in a plain class, not a React Component) which wraps a redux store that has a `react-redux-form` reducer in it.

2) The `Form` class exposes an API that is in some ways similar to `react-redux-form`, but we renamed methods that we didn't think were intuitively named. For instance, we have `resetModel` and `resetUi` instead of `reset` and `setInitial` (we think of the two parts of form state as "model state" and "ui state" so this worked well for the language we use internally). Also, all calls to `dispatch` are implementation details, so we're just calling something like `form.change('path.to.field', value)` instead of any consumer ever actually using `dispatch`. In our implementation, `redux` and `react-redux-form` have become implementation details. We were meticulous to make sure the abstraction didn't leak. There are a lot of reasons for this, but the number one reason for me was because I think I'll want to write a form's library using mobx for what I think will be huge perf benefits, and when I do that I want to retain our API.

_note: By doing these things we've basically come up with a "frameworkless"-ish form lib but we just happen to build it on top of `react-redux-form`_

3) We added a new method, `do`, which is a declarative API for doing effects on a form. `do` also batches changes for the perf benefits that will come with if you're making multiple changes. Our API for `Form#do` is like

```javascript
myForm.do({
  'some.field': {
    errors: { ... },
    pristine: false,
  },
  'some.other.field.somewhere': {
    value: 5, // this actually edits the model!
    valid: true,
  },
});
```

And again, under the hood `do` is just calling `react-redux-forms`'s actions inside a `batch`.

This Form class has enabled us to have form instances that we can pass around, without needing a global redux store, and without needing for a component to stay mounted on the UI for us to retain its state. For accessing state of a form throughout the application, we put together a `<FormRegistry/>` which uses React's context, sits at the top of our tree, and offers any part of the application access to any form.

Basic Example

```
<FormRegistry>
  <SomePage>
    <Main>
      <Form name='testName'>
        ...
      </Form>
    </Main>
    <SidePanel>
      <WithForm name='testName'>
        {form => (
          <span>The person's name is {form.get('firstName') + form.get('lastName')</span>
        )}
      </WithForm>
    </SidePanel>
  </SomePage>
</FormRegistry>
```

Our `<FormRegistry/>` has some other interesting features, but this is the crux of it.

Overall, we like keeping form state "local" (living in some form class) but we give access to that state via the context to other parts of the UI. This means we're not always dispatching to the app's global redux store (which for us presented perf problems after a while in a really large app).

Sorry it took me a while to reply. Hope this stuff makes sense. Happy to explain more if you're interested.
